great_expectations/datasource/fluent/batch_request.py

```diff
@@ -0,0 +1,225 @@
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING,
+    AbstractSet,
+    Any,
+    Callable,
+    Dict,
+    Mapping,
+    Optional,
+    Union,
+)
+
+import pydantic
+from pydantic import StrictStr
+from pydantic.json import pydantic_encoder
+from pydantic.schema import default_ref_template
+from typing_extensions import TypeAlias
+
+from great_expectations.core._docs_decorators import public_api
+from great_expectations.datasource.data_connector.batch_filter import (
+    BatchSlice,
+    parse_batch_slice,
+)
+
+if TYPE_CHECKING:
+    MappingIntStrAny = Mapping[Union[int, str], Any]
+    AbstractSetIntStr = AbstractSet[Union[int, str]]
+
+
+# BatchRequestOptions is a dict that is composed into a BatchRequest that specifies the
+# Batches one wants as returned. The keys represent dimensions one can filter the data along
+# and the values are the realized. If a value is None or unspecified, the batch_request
+# will capture all data along this dimension. For example, if we have a year and month
+# splitter, and we want to query all months in the year 2020, the batch request options
+# would look like:
+#   options = { "year": 2020 }
+BatchRequestOptions: TypeAlias = Dict[StrictStr, Any]
+
+
+@public_api
+class BatchRequest(pydantic.BaseModel):
+    """A BatchRequest is the way to specify which data Great Expectations will validate.
+
+    A Batch Request is provided to a Data Asset in order to create one or more Batches.
+
+    Args:
+        datasource_name: The name of the Datasource used to connect to the data.
+        data_asset_name: The name of the Data Asset used to connect to the data.
+        options: A dict that can be used to filter the batch groups associated with the Data Asset.
+            The dict structure depends on the asset type. The available keys for dict can be obtained by
+            calling DataAsset.batch_request_options.
+        batch_slice: A python slice that can be used to filter the sorted batches by index.
+            e.g. `batch_slice = "[-5:]"` will request only the last 5 batches after the options filter is applied.
+
+    Returns:
+        BatchRequest
+    """
+
+    datasource_name: StrictStr = pydantic.Field(
+        ...,
+        allow_mutation=False,
+        description="The name of the Datasource used to connect to the data.",
+    )
+    data_asset_name: StrictStr = pydantic.Field(
+        ...,
+        allow_mutation=False,
+        description="The name of the Data Asset used to connect to the data.",
+    )
+    options: BatchRequestOptions = pydantic.Field(
+        default_factory=dict,
+        allow_mutation=True,
+        description=(
+            "A map that can be used to filter the batch groups associated with the Data Asset. "
+            "The structure and types depends on the asset type."
+        ),
+    )
+    _batch_slice_input: Optional[BatchSlice] = pydantic.Field(
+        default=None,
+        allow_mutation=True,
+    )
```

Comment:
Can we just combine a pydantic `PrivateAttr` with traditional `@property` to deal with this, and then get rid of the stub and the all the special `Config` options?
https://docs.pydantic.dev/latest/usage/models/#private-model-attributes

Reply:
Can we make `batch_slice` a regular pydantic field and `validate_on_assignment`?

Reply:
This is a private attr (`underscore_attrs_are_private=True`). I'm not sure what other config options you are referring to, but we need a way to set the input of `batch_request.batch_slice` on `_batch_slice_input`.

Reply:
We need to save the `_batch_slice_input` to serialize it out in the `batch_slice` json property.

Reply:
Some of this pydantic configuration seems a bit overly complicated, I think it's probably worth taking another look to see what can be simplified. But I don't want to block on this, so I'm approving.

Reply:
>This is a private attr (underscore_attrs_are_private=True).


Elsewhere in the codebase we explicitly use `pydantic.PrivateAttr` to make fields private, so I think doing it via config like this could be confusing.

https://github.com/great-expectations/great_expectations/blob/b18f1af59add5c0479aa6f5dfc04082cbf3e66bf/great_expectations/datasource/fluent/interfaces.py#L265-L267


>but we need a way to set the input of `batch_request.batch_slice` on `_batch_slice_input`
I don't think we need this extra `_batch_slice_input` attribute at all. Can we just use a custom validator on `batch_slice` and do whatever needed parsing logic there?

Reply:
I reverted the Config `underscore_attrs_are_private` change to a `PrivateAttr` like you suggested. We used to need the alias until that approach no longer worked.

Reply:
The purpose of saving the input is so we can serialize it back out to config. We want the serialization/deserialization to be symmetrical, but if we lose the input upon deserialization we have to serialize something like `slice(start, stop, step)` back out to config.